include/flatbuffers/flatbuffers.h

```diff
@@ -616,19 +640,29 @@ class vector_downward {
 
   Allocator *allocator_;
   bool own_allocator_;
+  uoffset_t initial_size_;
   size_t reserved_;
   uint8_t *buf_;
   uint8_t *cur_;  // Points at location between empty (below) and used (above).
 
   void reallocate(size_t len) {
-    size_t old_reserved = reserved_;
+    assert(allocator_ != nullptr);
+    auto old_reserved = reserved_;
     auto old_size = size();
-    auto largest_align = AlignOf<largest_scalar_t>();
-    reserved_ += (std::max)(len, growth_policy(reserved_));
-    // Round up to avoid undefined behavior from unaligned loads and stores.
-    reserved_ = (reserved_ + (largest_align - 1)) & ~(largest_align - 1);
-    buf_ = allocator_->reallocate_downward(buf_, old_reserved, reserved_);
-    cur_ = buf_ + reserved_ - old_size;
+    size_t reserved;
+    if (old_reserved == 0) {
```

Comment:
auto reserved = old_reserved + (std::max)(len, old_reserved == 0 ? initial_size_ : growth_policy(old_reserved));
And why the cast? make initial_size_ a size_t?
Why both `reserved` and `reserved_`? harder to follow.
In fact, since we can already tell from`buf_` there is no buffer, we don't need `reserved_` to signal there's no buffer, and could have kept the older, simpler logic.